pkg/metrics/delivery_reporter.go

```diff
@@ -0,0 +1,167 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"knative.dev/pkg/metrics"
+	"strconv"
+	"time"
+)
+
+type DeliveryReporter struct {
+	podName               PodName
+	containerName         ContainerName
+	dispatchTimeInMsecM   *stats.Float64Measure
+	processingTimeInMsecM *stats.Float64Measure
+}
+
+type DeliveryReportArgs struct {
+	namespace  string
+	broker     string
+	trigger    string
+	filterType string
+}
+
+func (r *DeliveryReporter) register() error {
+	return view.Register(
+		&view.View{
+			Name:        "event_count",
+			Description: "Number of events received by a Trigger",
+			Measure:     r.dispatchTimeInMsecM,
+			Aggregation: view.Count(),
+			TagKeys: []tag.Key{
+				NamespaceNameKey,
+				BrokerNameKey,
+				TriggerNameKey,
+				TriggerFilterTypeKey,
+				ResponseCodeKey,
+				ResponseCodeClassKey,
+				PodNameKey,
+				ContainerNameKey,
+			},
+		},
+		&view.View{
+			Description: r.dispatchTimeInMsecM.Description(),
+			Measure:     r.dispatchTimeInMsecM,
+			Aggregation: view.Distribution(metrics.Buckets125(1, 10000)...), // 1, 2, 5, 10, 20, 50, 100, 1000, 5000, 10000
+			TagKeys: []tag.Key{
+				NamespaceNameKey,
+				BrokerNameKey,
+				TriggerNameKey,
+				TriggerFilterTypeKey,
+				ResponseCodeKey,
+				ResponseCodeClassKey,
+				PodNameKey,
+				ContainerNameKey,
+			},
+		},
+		&view.View{
+			Description: r.processingTimeInMsecM.Description(),
+			Measure:     r.processingTimeInMsecM,
+			Aggregation: view.Distribution(metrics.Buckets125(1, 10000)...), // 1, 2, 5, 10, 20, 50, 100, 1000, 5000, 10000
+			TagKeys: []tag.Key{
+				NamespaceNameKey,
+				BrokerNameKey,
+				TriggerNameKey,
+				TriggerFilterTypeKey,
+				PodNameKey,
+				ContainerNameKey,
+			},
+		},
+	)
+}
+
+// NewDeliveryReporter creates a new DeliveryReporter.
+func NewDeliveryReporter(podName PodName, containerName ContainerName) (*DeliveryReporter, error) {
+	r := &DeliveryReporter{
+		podName:       podName,
+		containerName: containerName,
+		// dispatchTimeInMsecM records the time spent dispatching an event to
+		// a Trigger subscriber, in milliseconds.
+		dispatchTimeInMsecM: stats.Float64(
+			"event_dispatch_latencies",
+			"The time spent dispatching an event to a Trigger subscriber",
+			stats.UnitMilliseconds,
+		),
+		// processingTimeInMsecM records the time spent between arrival at the Broker
+		// and the delivery to the Trigger subscriber.
+		processingTimeInMsecM: stats.Float64(
+			"event_processing_latencies",
+			"The time spent processing an event before it is dispatched to a Trigger subscriber",
+			stats.UnitMilliseconds,
+		),
```

Comment:
How would these be calculated? Via some custom attribute on the event?

Reply:
I have not looked to closely into integrating it yet, I'm going to do that as soon as I submit this PR but what do you mean by "some custom attribute on the event?"

Reply:
The ingress adds an event arrival time attribute to the event. I guess that's what Chen is asking?